lib/net/winrm/connection.rb

```diff
@@ -0,0 +1,272 @@
+            return nil unless response
+            return response.body if response.headers['Content-Type'].first =~ %r{\Aapplication\/soap\+xml}i
+            return '' if response.body.empty?
+
+            str = response.body.force_encoding('BINARY')
+            str.sub!(%r{^.*Content-Type: application\/octet-stream\r\n(.*)--Encrypted.*$}m, '\1')
+    
+            signature = str[4..19]
+            message = ntlm_client.session.unseal_message str[20..-1]
+            if ntlm_client.session.verify_signature(signature, message)
+              response.body = message
+              return
+            else
+              raise WinRMHTTPTransportError, 'Could not decrypt NTLM message.'
+            end
+          end
+
+          def ntlm_transform_request(ntlm_client, req)
+            return req if !req.opts['data']
+            req.opts['ctype'] = 'multipart/encrypted;protocol="application/HTTP-SPNEGO-session-encrypted";boundary="Encrypted Boundary"'
+            data = req.opts['data']
+            emessage = ntlm_client.session.seal_message data
+            signature = ntlm_client.session.sign_message data
+            edata = "\x10\x00\x00\x00#{signature}#{emessage}"
+
+            req.opts['data'] = body(edata,data.bytesize)
+            return req
+          end
+
+          def _send_request(message)
+            @mutex.synchronize {
+              opts = {
+                'uri' => self.uri,
+                'method' => 'POST',
+                'agent' => 'Microsoft WinRM Client',
+                'ctype' => 'application/soap+xml;charset=UTF-8',
+              }
+              if message
+                opts['data'] = message
+                opts['ntlm_transform_request'] = method(:ntlm_transform_request)
+                opts['ntlm_transform_response'] = method(:ntlm_transform_response)
+              end
+              request = self.http_client.request_cgi(opts)
+              response = self.http_client.send_recv(request,self.timeout,true)
+              if response
+                WinRM::ResponseHandler.new(response.body, response.code).parse_to_xml
+              else
+                raise WinRM::WinRMHTTPTransportError.new("No response")
+              end
+            }
+          end
+
+          def send_request(message)
+            unless self.first_request_sent
+              _send_request(nil)
+              self.first_request_sent = true
+            end
+            _send_request(message)
+          end
+          protected
+            attr_accessor :http_client, :uri, :timeout
+
+            # Need to send an empty first request to potentially set up an encryption
+            # channel - required if allowUnencrypted is set to false, which is the case
+            # by default
+            attr_accessor :first_request_sent
+        end
+    
+        def create_transport(connection_opts)
+          raise NotImplementedError unless connection_opts[:transport] == :rex
```

Comment:
Could we please rename our transport to `:rex_http`? That would more closely align with the `RexHttpTransport` class name and put us in a better position if we got around to adding SSH as a transport.

Reply:
I've made the change; however I'm not sure what you mean about an SSH transport. The intention of `connection_opts[:transport]` is to tell the WinRM library which Transport subclass to instantiate: Plaintext? Negotiate? Kerberos? CredSSP? We want none of these built-in transports, because they all use HTTPClient. Instead, we want our own class: `RexHttpTransport`. I don't think SSH is relevant here. Let me know if you think the code needs additional clarification in there.